### ISSUE TITLE
Back modifier

### ISSUE THREAD

**Issue by andreysamode** [ypsu/myts#2]
I just noticed that the Back key modifier doesn't seem to be working on Kindle 3. On myts-8, I can press Back+v to make "/", but on this version it simply prints "v". Am I missing something or is this a bug?

Thanks,
Andrey

**Comment by ypsu** [comment]
I really haven't touched this repo for 3 years and even then it was just for storing my personal settings (like the different font). Now I see it even lacks all the scripts necessary to install this. :)

The key combo you mentioned works on my kindle though. It seems I modified keydefs.ini in f7d56820be534531b2d0436599e3cc843c1639df. Have you copied that file over to your kindle too? If yes, does it work if you use the prebuilt binary? That's what I'm using.

**Comment by andreysamode** [comment]
I copied the whole thing over, replacing myts-8. What do you mean by prebuilt binary?

P.S. If you don't feel like looking into this too deeply, no worries. I'm perfectly happy to keep using Myts 8 and using your commit log to fix a few things myself.

**Comment by ypsu** [comment]
Ahhh, I think I'm just stupid. In keydefs.ini comment the kindle dx keys and uncomment the kindle 3 key definitions and then try again. I probably installed this on my dx after my 3 and that's why I left it in that state.

**Comment by andreysamode** [comment]
Yup, that was it. It works beautifully now :-)

By the way, how come the myts binary filesize is 281KB whereas the original was only 31KB?

And another question if you don't mind: is it at all possible to change the caret from a block to a pipe "|" or an underscore "_"?

Thanks,
Andrey.

**Comment by ypsu** [comment]
re size: Two things: I forgot to run strip on the binary. When I run it (i.e. `strip myts`), I get it down to 102K. The other bit is that I'm linking statically while the original links dynamically. So my build contains the libc too. You can see this if you run `ldd myts`. IIRC I started the static building because the build from my machine requires a very recent glibc so I couldn't get the binary started on the kindle. On the other hand musl libc didn't have such problems.

re caret: It is certainly doable albeit a bit tricky. The caret seems to be rendered in launchpad.c's print_buf function. This is the relevant bit:

```c
            if ( i == cur )
                bg |= 0x88;
            get_char_pixmap(lps->fb->font, cc, &char_pixmap) ;
            pix_blt(&lps->fb->pixmap, x, y,
                &char_pixmap, 0, 0, -1, -1, bg) ;
```

It is implemented by changing the background color of the underlying glyph. That background color covers the whole grid of the glyph. To change you'll probably need to change the pix_blt function to deal with the caret rendering. E.g. you add a boolean flag called is_caret which if true, then it blits a solid line to the bottom. Or better, add a new function to render horizontal / vertical lines and just invoke that. Or add an option to pix_blt to bitwise or the glyph onto framebuffer instead of memcpy and use that to render the caret after the main glyph -- then you can use arbitrary glyph for the caret. Or something like that. Let me know if unclear and then I dig deeper to see if I can come up with something.

**Comment by ypsu** [comment]
re size: I now committed 136525779bfb0990ef81efb7db2bde68826afdce which reduced the prebuilt binary to 74K.

**Comment by andreysamode** [comment]
The new reduced size binary works wonderfully.

Thanks for the cursor instructions. I tried to alter pix_blt in launchpad.c, but none of the changes took. I suppose that's because I have to re-compile it? Which versions of libgc, etc do I need for that? If this is too much trouble, I'm happy with the block caret—it's only a slight preference, so no worries.

The reason why I'm investigating all of this is because I'm experimenting with connecting a PS/2 keyboard to my Kindle 3. I've got the PS/2 keyboard connected to Teensy++ and 20 pins via a ribbon cable going into kindle's own keyboard port. So far, when I press a key and activate a pin, I get a bunch of gibberish (for example: 6ih7v on pin 11, or [home]2tdz on pin 8). The good news is that the gibberish is consistent, so I wonder whether it would be possible to re-interpret these sets of characters into letters. For example, activating pin8 and pin11 could be rewritten as say "Q". I know this is well beyond the scope of this project, but if you have any thoughts on the matter, I would much appreciate it.

**Comment by ypsu** [comment]
re cursor: Yes, you need to recompile. The kindle's architecture is arm6 and is running linux while your computer's is probably an intel architecture (x86_64) so compiling with the default settings on that machine won't work. I recommend setting up a raspberry pi (it's enough to do this on a virtual machine), install the musl and gcc packages on it. Then anything you build statically via musl-gcc, you can just simply copy over the kindle and run it.

re keyboard: Although what you wrote is a bit over my head (I don't know much about kindle internals or about electronics in general), I don't see big problems with the approach you envisioned. Perhaps kindle might not support all the keys if you do it this way since the kindle keyboard has very few keys. An alternative would be to use the usb port to connect the keyboard to (not entirely sure how, but surely it's doable). Wouldn't that be simpler?

By the way, what do you want to do? When I played with this about 3 years ago, I wanted to use the kindle as a computer monitor. I usually work from a tmux session running in a single terminal, so I thought a terminal on a kindle would be enough. I used the dx for this since it had a bigger screen. I ssh'd to my primary machine from the kindle and attached to a tmux session. Tmux was the primary reason I forked the myts project and fixed a few bugs in it. It still has some, but it is now usable for light sessions. The keyboard was attached to the primary computer, the kindle just acted as a screen. Then I turned off the primary monitor, and hacked away on the kindle screen. It has slow refresh rate though. One needs to change their computer habits into a way which minimizes changes on the screen (scrolling and jumping around is bad) and doesn't require instant feedback (e.g. you have to type blindly). It's fun though. I stopped the experiment eventually because for one I often had to switch the primary monitor back on when I wanted to visit a website which only worked from a modern browser, and second I was very concerned that it would the kill the screen fast. IIUC the screen has limited refresh cycles, and this usecase was depleting it way too quickly. I might decide to experiment with this again if I get bored since I'm not using the dx these days anyways.

**Comment by andreysamode** [comment]
I see. That makes sense about the architecture. Thanks. Could it be compiled on the Kindle itself or would that require too complex of a setup?

I'm attempting to create an eInk typewriter for myself. I write a lot and notice a huge difference in quality of writing when I stare at a monitor vs a kindle (I currently use AlphaSmart Neo, but it has limited storage). All I really need is a working terminal (which I have, thank to you :-) and Vim (which also works perfectly fine).

I'm fairly new this this myself. Connecting via USB was the first avenue I investigated, but unfortunately it doesn't seem like anyone has been successful—apparently, you need USB Host functionality, which Kindles don't have. It's been done on a Nook, but the battery drain from USB host is so intense, it doesn't seem too viable. Ultimately, I'd love to get a larger power bank, hook it up to a solar panel on the back and have a solar eInk typewriter. That part should't be too difficult if the basic concept of connecting a keyboard to the kindle works.

The original inspiration came from the Frankenkindle project:
http://breadboardconfessions.blogspot.com/2011/08/frankenkindle-prototype-demo.html
But instead of just a few navigation buttons, I'd like to connect the entire keyboard. At the moment I have a PS/2 keyboard going into Teensy++, then 20 pins going out to Kindle's own keyboard port. Kindle's own keyboard is a simple row-column matrix of switches. Here's the map if you're curious:
http://breadboardconfessions.blogspot.com/2011/06/by-request-kindle-3-keyboard-map.html
Basically to trigger a key you need to cross a row pin with a column pin. So crossing pin 2 with pin 12 should produce "q".

I didn't think it would even work without a multiplexer (switchboard to control which pins get crossed with which), but somehow the five-way controller works totally fine just by activating the pins themselves (Pins 15-19). They're supposed to only activate when they are crossed with Pin 20, but even if I disconnect that pin, they still work. It's a bit of an electrical puzzle for me at the moment.

**Comment by ypsu** [comment]
Compiling on kindle is certainly doable but quite involved to set it up. I still recommend setting up a raspberry pi instead. Just search for "qemu raspberry" to find instructions how to set up a virtual machine for it. Once booted run `sudo apt-get install gcc musl-tools`. Then compile a hello world app via `musl-gcc -mcpu=arm6 -static hello.c`, copy it over to the kindle and if it works, you should be able to compile myts too. Alternatively, if you want to keep using your primary machine (probably more convenient), look into how to do arm crosscompiling. It's quite straightforward too I just never done it because I happen to own an rpi. :)

The teensy you mentioned is fully programmable, right? So even if you can only make the 4 directional buttons work, you can still simulate a full keyboard by it. You configure your teensy to emit 4 kindle keypresses for each real keypress. Then you would run a software on the kindle to translate those codes into simulated kindle keypresses. 4 * 2bit = 8bit so you could cover all the keys with that.

Do you write near your primary computer or smartphone? You could connect a bluetooth keyboard to your computer, then connect the kindle to your computer over wifi and ssh from the kindle to your computer like I did. Although probably the wifi is also draining the battery a lot so not ideal. On the other hand you could plug the kindle onto a small portable battery pack and it would last quite long.

By the way, looking at the AlphaSmart Neo, it seems it has quite a small screen. You are probably not editing text on it, you just write. Suppose that you can touch type, do you even need a screen? Those teensies have 256k flash memory. Is it possible to write to that from the teensy directly? The full Gulliver's Travels fits into that if I compress it with the simplest lzw compression (I actually downloaded it from gutenberg and ran ncompress on it -- I got 233K). If not, you can always attach a small sdcard to it. So if you don't need a screen, you can just type right into the keyboard's controller! Basically you queue up the keypresses into the flash and when plug the keyboard to the computer, you drain the queue. This wouldn't require lot of battery, I believe. But you would look like a weirdo when you are typing. Or you could even use pen and paper. You can then train some ocr software to recognize your handwriting and digitize your notes whenever you get near a computer. The only reason I'm telling you this is because at this point I'm seeing too many computers everywhere, and now I'm trying to get people to consider older techs for old problems -- to have less computers, more present people. Nevertheless, the above technical challenge still sounds fun. :)

**Comment by andreysamode** [comment]
Hah! I love it. A keyboard without a screen might be worth just for the weird looks :-)

I'm working on a novel, so I have to go over the chapters and edit quite a bit. Surprisingly, the Neo screen isn't too small for me. In AlphaWord, there's a way to auto-fold the text based on the number of paragraph breaks. Between that and search functionality, editing is actually pretty doable. The main issue is that I can't fit the whole novel onto the Neo, so I have to shuffle files on and off the computer.

I share your sentiment about older tech and presence. I find that a lot of the newer tech has a "dumbing" effect on my consciousness. I love the Neo precisely for its simplicity. Pen and paper isn't really viable for my writing because I have to edit quite a bit, but I do like using them to visually sketch out ideas.

I just ordered a couple of multiplexers. They're basically tiny analog switchboards that can connect the control signal (A0, B0) to any of the pins (A1-8 on one and B1-16 on the other). I think by wiring the control pins (A0, B0) to each other, and connecting row pins to A1-8 and column pins to B1-16, I'll be able to replicate kindle's keyboard functionality.

I installed an Ubuntu virtual machine and inside it, a Linaro ARM virtual environment with musl-tools. When I try to compile a hello world app with `musl-gcc -mcpu=arm6 -static hello.c`, I get:
`
hello.c:1:0: warning: target CPU does not support THUMB instructions [enabled by default]
 /* Hello World program */
 ^
/tmp/cc29yFUC.s: Assembler messages:
/tmp/cc29yFUC.s:35: Error: selected processor does not support ARM mode `bx lr'
`

From what I understand, "thumb" is a newer technology, unavailable on older Arm CPUs. To disable, I compile with `musl-gcc -mcpu=arm6 -mno-thumb-interwork -static hello.c`, which seems to work. However, when I try to run it on the kindle, I get "Illegal instruction".

**Comment by ypsu** [comment]
On my raspberrypi it works without any flags so probably on a different machine gcc needs a few more flags than the one I posted. Try all the -m* flags from here: https://github.com/skyleecm/vim-for-kindle. It claims to work for vim on kindle 3 so probably does for your hello world too. If it still doesn't work, I'll try to set up a similar environment to yours and see if I can make it work.

**Comment by andreysamode** [comment]
Hurray! I got the Hello World app to work on the kindle :-)

Here are the steps I took for creating the Kindle cross-compile environment:
* Install Ubuntu mini
* sudo apt-get install scratchbox2
* wget http://www.scratchbox.org/download/files/sbox-releases/apophis/tarball/scratchbox-toolchain-cs2007q3-glibc2.5-arm7-1.0.14-2-i386.tar.gz
* tar xvfz scratchbox-toolchain-cs2007q3-glibc2.5-arm7-1.0.14-2-i386.tar.gz
* (This next armv7 line is probably unnecessary, but I'll keep it here for reference.) sudo sb2-init --host -c /usr/bin/qemu-arm armv7 /home/username/scratchbox/compilers/cs2007q3-glibc2.5-arm7/bin/arm-none-linux-gnueabi-gcc
* sudo sb2-init --host -c /usr/bin/qemu-arm armv6 /home/username/scratchbox/compilers/cs2007q3-glibc2.5-arm7/bin/arm-none-linux-gnueabi-gcc
* sudo sb2 sb2-build-libtool --host
* export SBOX_EXTRA_COMPILER_ARGS="-march=armv6j -mcpu=arm1136jf-s -mtune=arm1136jf-s -mfpu=vfp"
* sb2 -t armv6 gcc -static hello.c -o hello

Now when I try to compile myts with `sb2 -t armv6 gcc -static myts.c -o myts`, I get:
`/tmp/ccMsmIkf.o:(.data+0x0): undefined reference to 'lpad'
collect2: ld returned 1 exit status`

__UPDATE (3)__
To replicate your particular environment, I did the following steps:
* mkdir proj
* mkdir proj/oss
* cd proj/oss
* wget https://www.musl-libc.org/releases/musl-0.9.14.tar.gz
* tar xvfz musl-0.9.14.tar.gz
* cd musl-0.9.14
* sb2 -t armv6 ./configure
* sb2 -t armv6 make

I also downloaded Kindle Cross-Compile Toolchain to use instead of Ubuntu's /usr/include:
https://github.com/samsheff/Amazon-Kindle-Cross-Toolchain
and changed Makefile line:
`CFLAGS += -nostdinc -isystem ~/proj/oss/musl-0.9.14/include/ -isystem /usr/include`
to
`CFLAGS += -nostdinc -isystem ~/proj/oss/musl-0.9.14/include/ -isystem ~/kindle-crosscompile-toolchain/arm-kindle-linux-gnueabi/arm-kindle-linux-gnueabi/sysroot/usr/include/`

Now, in the myts directory, when I run:
`sb2 -t armv6 make`

I get the following error:
`gcc -static -Os -Wall -Werror -nostdinc -isystem ~/proj/oss/musl-0.9.14/include/ -isystem ~/kindle-crosscompile-toolchain/arm-kindle-linux-gnueabi/arm-kindle-linux-gnueabi/sysroot/usr/include/ -I. -DNODEBUG   -c -o launchpad.o launchpad.c
cc1: warnings being treated as errors
launchpad.c: In function 'launchpad_init':
launchpad.c:507: warning: dereferencing type-punned pointer will break strict-aliasing rules
launchpad.c:509: warning: dereferencing type-punned pointer will break strict-aliasing rules
make: *** [launchpad.o] Error 1`

If I alter Makefile and remove ` -Werror`, it compiles successfully, doesn't give "segmentation fault" on the Kindle if I run it via USB Networking and works via Launchpad!

Hurray!

Is that error anything to worry about?

**Comment by ypsu** [comment]
Whoa, you've set up a full cross compile environment. In that case you shouldn't have needed to duplicate my environment, hah. If you want to simplify, you can change `CC=musl-gcc` back to `CC=gcc` and just completely remove the `CFLAGS += -isystem /usr/lib/musl/include -isystem /usr/include` line. I've removed the proj/oss bits shortly after you opened this bug, you wouldn't have seen it if you've done a `git pull`. If it still doesn't work, you could always try the original makefile from the original package. However since you made everything work, maybe it doesn't really need further tweaking for now. :)

That warning is questionable. I cannot get gcc print it to me and I've tried it with gcc 4.8 and 6.3. In this particular case it should be harmless but if you are concerned you can add `-fno-strict-aliasing` to the cflags to disable optimizations which break when the code doesn't adhere to the strict aliasing rules. But really, you can ignore it. There are bunch of other warnings (you can see them via `-Wextra`) which ought to be fixed if anybody were to care about this project. :)

**Comment by andreysamode** [comment]
Sounds good. I just cracked my kindle screen while taking it apart, so I'm putting the project on hold until the new screen arrives. I'll post a little update once I get things working. Thanks for all the help and super enjoyable correspondence :-)

**Comment by andreysamode** [comment]
Just to close this out, I did manage to connect a PS/2 keyboard through an Arduino and two multiplexers back to back into the Kindle Keyboard's keyboard slot and got it to type on a chrooted Debian Emacs. It looked really hackish and would require quite a bit more work to fit it into a practical form, so I decided the project was not worth all the effort. Especially, since an actual eInk laptop is in the works by Onyx Boox: https://www.youtube.com/watch?v=ni2czAM2NDU

Thanks for the help. It was a fun and challenging project, even though it turned out more of a proof-of-concept than a practical writing machine. I'm back to happily writing on my Neo for now :-)
